File: strategies/donchian-sqeeze-momentum-p1/donchian-sqeeze-momentum-p1.ts
import { createChart, LineStyle, IChartApi, ISeriesApi, SeriesMarker, Time } from 'lightweight-charts';

interface StrategyResult {
	basis: number[];
	upper: number[];
	lower: number[];
	upper_vol: number[];
	lower_vol: number[];
	upper_inner: number[];
	lower_inner: number[];
	fu: number[];
	fl: number[];
	strong_long: number[];
	strong_short: number[];
	weak_long: number[];
	weak_short: number[];
	trend: number[];
	bullish_shift: number[];
	bearish_shift: number[];
	timestamps: number[];
	visualization: {
		colors: {
			up: string;
			down: string;
			neutral: string;
			basis_up: string;
			basis_down: string;
			cloud_up: string;
			cloud_down: string;
			cloud_neutral: string;
		};
		styles: {
			basis: { lineWidth: number };
			bands: { lineWidth: number; style: string };
			volatility: { lineWidth: number; style: string };
			signals: { size: string };
		};
	};
}

export class DonchianSqueezeMomentumChart {
	private chart: IChartApi;
	private candlestickSeries: ISeriesApi<"Candlestick">;
	private basisSeries: ISeriesApi<"Line">;
	private upperSeries: ISeriesApi<"Line">;
	private lowerSeries: ISeriesApi<"Line">;
	private upperVolSeries: ISeriesApi<"Line">;
	private lowerVolSeries: ISeriesApi<"Line">;
	private upperInnerSeries: ISeriesApi<"Line">;
	private lowerInnerSeries: ISeriesApi<"Line">;
	private fuSeries: ISeriesApi<"Line">;
	private flSeries: ISeriesApi<"Line">;
	private markerSeries: ISeriesApi<"Line">;

	constructor(container: HTMLElement) {
		this.chart = createChart(container, {
			layout: {
				background: { color: '#ffffff' },
				textColor: '#333',
			},
			grid: {
				vertLines: { color: '#f0f0f0' },
				horzLines: { color: '#f0f0f0' },
			},
			
			rightPriceScale: {
				borderColor: '#d1d4dc',
				scaleMargins: {
					top: 0.1,
					bottom: 0.1,
				},
			},
			timeScale: {
				borderColor: '#d1d4dc',
				timeVisible: true,
			
			},
		});

		// Initialize all series
		this.candlestickSeries = this.chart.createCandlestickSeries({
			upColor: '#26a69a',
			downColor: '#ef5350',
			borderVisible: false,
			wickUpColor: '#26a69a',
			wickDownColor: '#ef5350',
		});

		this.basisSeries = this.chart.addLineSeries({
			title: 'Basis',
			color: '#2196F3',
			lineWidth: 2,
		});

		this.upperSeries = this.chart.addLineSeries({
			title: 'Upper Band',
			color: '#FF4081',
			lineWidth: 1,
			lineStyle: LineStyle.Dashed,
		});

		this.lowerSeries = this.chart.addLineSeries({
			title: 'Lower Band',
			color: '#FF4081',
			lineWidth: 1,
			lineStyle: LineStyle.Dashed,
		});

		this.upperVolSeries = this.chart.addLineSeries({
			title: 'Upper Volatility',
			color: '#9C27B0',
			lineWidth: 1,
			lineStyle: LineStyle.Dotted,
		});

		this.lowerVolSeries = this.chart.addLineSeries({
			title: 'Lower Volatility',
			color: '#9C27B0',
			lineWidth: 1,
			lineStyle: LineStyle.Dotted,
		});

		this.upperInnerSeries = this.chart.addLineSeries({
			title: 'Upper Inner',
			color: '#7B1FA2',
			lineWidth: 1,
			lineStyle: LineStyle.Dotted,
		});

		this.lowerInnerSeries = this.chart.addLineSeries({
			title: 'Lower Inner',
			color: '#7B1FA2',
			lineWidth: 1,
			lineStyle: LineStyle.Dotted,
		});

		this.fuSeries = this.chart.addLineSeries({
			title: 'FU Line',
			color: '#4CAF50',
			lineWidth: 1,
		});

		this.flSeries = this.chart.addLineSeries({
			title: 'FL Line',
			color: '#F44336',
			lineWidth: 1,
		});

		this.markerSeries = this.chart.addLineSeries({
			title: 'Signals',
			lineWidth: 0,
		});
	}

	public async updateChart(symbol: string, timeframe: string): Promise<void> {
		try {
			const response = await fetch(`/api/donchian-sqeeze-momentum-p1?symbol=${symbol}&timeframe=${timeframe}`);
			if (!response.ok) {
				throw new Error('Failed to fetch data');
			}
			const result: StrategyResult = await response.json();

			// Update all series
			const seriesData = this.prepareSeriesData(result);
			
			// Set data for each series
			this.basisSeries.setData(seriesData.basis);
			this.upperSeries.setData(seriesData.upper);
			this.lowerSeries.setData(seriesData.lower);
			this.upperVolSeries.setData(seriesData.upperVol);
			this.lowerVolSeries.setData(seriesData.lowerVol);
			this.upperInnerSeries.setData(seriesData.upperInner);
			this.lowerInnerSeries.setData(seriesData.lowerInner);
			this.fuSeries.setData(seriesData.fu);
			this.flSeries.setData(seriesData.fl);

			// Set markers for signals
			this.markerSeries.setMarkers(this.prepareMarkers(result));

			// Fit content
			this.chart.timeScale().fitContent();

		} catch (error) {
			console.error('Error updating chart:', error);
		}
	}

	private prepareSeriesData(result: StrategyResult) {
		const mapToTimeValue = (values: number[]) => 
			values.map((value, index) => ({
				time: result.timestamps[index],
				value: value,
			}));

		return {
			basis: mapToTimeValue(result.basis),
			upper: mapToTimeValue(result.upper),
			lower: mapToTimeValue(result.lower),
			upperVol: mapToTimeValue(result.upper_vol),
			lowerVol: mapToTimeValue(result.lower_vol),
			upperInner: mapToTimeValue(result.upper_inner),
			lowerInner: mapToTimeValue(result.lower_inner),
			fu: mapToTimeValue(result.fu),
			fl: mapToTimeValue(result.fl),
		};
	}

	private prepareMarkers(result: StrategyResult): SeriesMarker<Time>[] {
		const markers: SeriesMarker<Time>[] = [];

		for (let i = 0; i < result.timestamps.length; i++) {
			if (result.strong_long[i]) {
				markers.push({
					time: result.timestamps[i],
					position: 'belowBar',
					color: result.visualization.colors.up,
					shape: 'arrowUp',
					text: 'Strong Long',
				});
			} else if (result.strong_short[i]) {
				markers.push({
					time: result.timestamps[i],
					position: 'aboveBar',
					color: result.visualization.colors.down,
					shape: 'arrowDown',
					text: 'Strong Short',
				});
			} else if (result.weak_long[i]) {
				markers.push({
					time: result.timestamps[i],
					position: 'belowBar',
					color: result.visualization.colors.neutral,
					shape: 'circle',
					text: 'Weak Long',
				});
			} else if (result.weak_short[i]) {
				markers.push({
					time: result.timestamps[i],
					position: 'aboveBar',
					color: result.visualization.colors.neutral,
					shape: 'circle',
					text: 'Weak Short',
				});
			}
		}

		return markers;
	}

	public resize(width: number, height: number): void {
		this.chart.resize(width, height);
	}

	public remove(): void {
		this.chart.remove();
	}
}

File: strategies/donchian-sqeeze-momentum-p1/donchian-sqeeze-momentum-p1.py
import numpy as np
import pandas as pd
import requests
import asyncio
import sys
import json
from typing import Dict, List, Any

class DonchianSqueezeMomentumStrategy:
    def __init__(self):
        # Strategy Parameters from PineScript
        self.length = 50
        self.mult = 3.0
        self.atr_period = 14
        self.con = 1
        self.colors = {
            'up': '#00ffbb',
            'down': '#ff1100',
            'neutral': '#808080',
            'basis_up': '#00ffbb',
            'basis_down': '#ff1100',
            'cloud_up': 'rgba(0, 255, 187, 0.3)',
            'cloud_down': 'rgba(255, 17, 0, 0.3)',
            'cloud_neutral': 'rgba(128, 128, 128, 0.3)'
        }

    async def fetch_candle_data(self, symbol: str, timeframe: str) -> Dict[str, List[float]]:
        """Fetch candle data from the API"""
        try:
            response = requests.get(
                'http://localhost:5000/fetch_candles',
                params={'symbol': symbol, 'timeframe': timeframe}
            )
            response.raise_for_status()
            data = response.json()
            
            return {
                'open': [candle['open'] for candle in data],
                'high': [candle['high'] for candle in data],
                'low': [candle['low'] for candle in data],
                'close': [candle['close'] for candle in data],
                'volume': [candle.get('volume', 0) for candle in data],
                'time': [candle['time'] for candle in data]
            }
        except Exception as e:
            print(f"Error fetching candle data: {e}")
            return {'open': [], 'high': [], 'low': [], 'close': [], 'volume': [], 'time': []}

    def calculate_donchian_channels(self, highs: np.ndarray, lows: np.ndarray) -> tuple:
        """Calculate Donchian Channel components"""
        highest = pd.Series(highs).rolling(window=self.length).max().values
        lowest = pd.Series(lows).rolling(window=self.length).min().values
        basis = (highest + lowest) / 2
        return highest, lowest, basis

    def calculate_atr(self, high: np.ndarray, low: np.ndarray, close: np.ndarray) -> np.ndarray:
        """Calculate Average True Range"""
        high_low = high - low
        high_close = np.abs(high - np.roll(close, 1))
        low_close = np.abs(low - np.roll(close, 1))
        
        ranges = np.vstack([high_low, high_close, low_close])
        true_range = np.max(ranges, axis=0)
        true_range[0] = ranges[0, 0]
        
        return pd.Series(true_range).rolling(window=self.atr_period).mean().values

    def calculate_standard_deviation(self, close: np.ndarray) -> np.ndarray:
        """Calculate Standard Deviation"""
        return pd.Series(close).rolling(window=self.length).std().values

    def calculate_crossovers(self, series1: np.ndarray, series2: np.ndarray) -> tuple:
        """Calculate crossover and crossunder signals"""
        crossover = (series1[:-1] <= series2[:-1]) & (series1[1:] > series2[1:])
        crossunder = (series1[:-1] >= series2[:-1]) & (series1[1:] < series2[1:])
        return np.append(crossover, False), np.append(crossunder, False)

    def calculate_strategy(self, data: Dict[str, List[float]]) -> Dict[str, Any]:
        """Main calculation function"""
        closes = np.array(data['close'])
        highs = np.array(data['high'])
        lows = np.array(data['low'])
        timestamps = np.array(data['time'])

        # Calculate Donchian Channel components
        highest, lowest, basis = self.calculate_donchian_channels(highs, lows)

        # Calculate Standard Deviation and ATR
        stdev = self.calculate_standard_deviation(closes)
        dev = stdev * self.mult
        vol = self.calculate_atr(highs, lows, closes)

        # Initialize arrays for fbasis and range
        fbasis = np.zeros_like(closes)
        range_values = np.zeros_like(closes)

        # Calculate fbasis and range with memory
        for i in range(len(closes)):
            if i == 0:
                fbasis[i] = basis[i]
                range_values[i] = dev[i]
            else:
                fbasis[i] = fbasis[i-1] if basis[i] == basis[i-self.con] else basis[i]
                range_values[i] = range_values[i-1] if fbasis[i] == fbasis[i-1] else dev[i]

        # Calculate all bands
        upper = fbasis + range_values
        lower = fbasis - range_values
        fu = fbasis + vol
        fl = fbasis - vol
        uu = upper + vol
        ul = upper - vol
        lu = lower + vol
        ll = lower - vol

        # Calculate crossover signals
        strong_long = np.zeros_like(closes, dtype=bool)
        strong_short = np.zeros_like(closes, dtype=bool)
        weak_long = np.zeros_like(closes, dtype=bool)
        weak_short = np.zeros_like(closes, dtype=bool)

        for i in range(1, len(closes)):
            strong_long[i] = closes[i-1] >= uu[i-1] and closes[i] < uu[i]
            strong_short[i] = closes[i-1] <= ll[i-1] and closes[i] > ll[i]
            weak_long[i] = (closes[i-1] >= ul[i-1] and closes[i] < ul[i]) and not strong_long[i]
            weak_short[i] = (closes[i-1] <= lu[i-1] and closes[i] > lu[i]) and not strong_short[i]

        # Calculate trend
        trend = np.where(closes > fbasis, 1, -1)

        # Calculate additional signals for plotting
        bullish_shift, bearish_shift = self.calculate_crossovers(closes, fbasis)

        return {
            'basis': fbasis.tolist(),
            'upper': upper.tolist(),
            'lower': lower.tolist(),
            'upper_vol': uu.tolist(),
            'lower_vol': ll.tolist(),
            'upper_inner': ul.tolist(),
            'lower_inner': lu.tolist(),
            'fu': fu.tolist(),
            'fl': fl.tolist(),
            'strong_long': strong_long.tolist(),
            'strong_short': strong_short.tolist(),
            'weak_long': weak_long.tolist(),
            'weak_short': weak_short.tolist(),
            'trend': trend.tolist(),
            'bullish_shift': bullish_shift.tolist(),
            'bearish_shift': bearish_shift.tolist(),
            'timestamps': timestamps.tolist(),
            'visualization': {
                'colors': self.colors,
                'styles': {
                    'basis': {'lineWidth': 2},
                    'bands': {'lineWidth': 1, 'style': 'dashed'},
                    'volatility': {'lineWidth': 1, 'style': 'dotted'},
                    'signals': {'size': 'tiny'}
                }
            }
        }

async def main():
    if len(sys.argv) != 3:
        print("Usage: python script.py <symbol> <timeframe>")
        sys.exit(1)

    symbol = sys.argv[1]
    timeframe = sys.argv[2]

    strategy = DonchianSqueezeMomentumStrategy()
    
    # Fetch candle data
    candle_data = await strategy.fetch_candle_data(symbol, timeframe)
    
    if not candle_data['close']:
        print(json.dumps({"error": "No data available"}))
        sys.exit(1)

    # Calculate strategy results
    results = strategy.calculate_strategy(candle_data)
    
    # Print results as JSON
    print(json.dumps(results))

if __name__ == "__main__":
    asyncio.run(main())